bpf: nat: Adding a new config field to enable SNAT for remote node. #35823

behzad-mir opened the pull request:
- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [X] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [X] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [X] Provide a title or release-note blurb suitable for the release notes.
- [X] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [X] Thanks for contributing!

<!-- Description of change -->
Following the issues discussed in [35823](https://github.com/cilium/cilium/issues/35823) and [17177](https://github.com/cilium/cilium/issues/17177) we are adding a new config field "enable-remote-node-snat" to enable SNAT for remote node in BPF routing mode.

Currently, Cilium ebpf based masquerading does not perform SNAT on packets destined for a remote node, even if the node CIDR is not listed in the non-masquerade list. This behavior differs from the upstream ip-masq-agent. This behavior prevents pods-remote nodes communication in a BPF-masquerade enabled cluster when pod and node network are in different subnets.

With this change if the new confimap field "enable-remote-node-snat" is set to "true"
then packets destined to remote nodes will be SNATed in bpf routing mode.

Changes are reflected in bpf/lib/nat.h :

Before:
```c
if (identity_is_remote_node(remote_ep->sec_identity))
     return NAT_PUNT_TO_STACK;
``` 
After:
```c
if (identity_is_remote_node(remote_ep->sec_identity)) {
    #ifdef ENABLE_REMOTE_NODE_SNAT
    {
	    target->addr = IPV4_MASQUERADE;
	    return NAT_NEEDED;
    }
    #endif
    return NAT_PUNT_TO_STACK;
}
``` 

In order to verify and test the expected behavior, we followed the repro in [35823](https://github.com/cilium/cilium/issues/35823#issuecomment-2607873964) in a cilium bpf routing cluster with the addition of enable-remote-node-snat = "true" in the configmap. ICMP packet from a pod to a remote node initiated.
The TCP-dump from remote node indicates that the source IPs for incoming packet SNATed to the ip of source node.


Fixes: #35823

```release-note
Add a new config field to enable remote node masquerading in BPF routing mode. This can help to establish the pods-remote nodes communication in a BPF-masquerade enabled cluster when pod and node network are in different subnets
```

pchaigno reviewed (dismissed):
Could you add the additional context in the commit description and not the PR description?

> In order to verify and test ithe change. We followed the repro in https://github.com/cilium/cilium/issues/35823 and verified that the source address is getting SNATed to the source node IP.

Can you describe exactly how you did that with the current code? It does not work for me.

pchaigno commented on daemon/cmd/daemon_main.go:
```diff
@@ -661,6 +661,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	option.BindEnv(vp, option.EnableIPv4Masquerade)
 
+	flags.Bool(option.EnableRemoteNodeSnat, false, "SNAT IPv4 packets from endpoints leaving the host destined to a remote node")
+	option.BindEnv(vp, option.EnableRemoteNodeSnat)
```
Could you also document the new flag under https://docs.cilium.io/en/latest/operations/upgrade/#upgrade-notes?

pchaigno replied:
For this, you'll probably want to mention that the behavior from this flag isn't exactly the optimal behavior discussed in https://github.com/cilium/cilium/issues/17177 because packets going to the nodes' InternalIP addresses will also be masqueraded (unless I missed some other check we have).

AFAIU, to implement the optimal behavior, we'd need to tell our datapath which type of IP address each node IP address is. That could probably be achieved by extending the node ID map.

behzad-mir replied:
Understood. Will adding this

pchaigno replied:
This wasn't addressed :)

behzad-mir replied:
Thanks @pchaigno, I missed this one. I added a section as you have requested. Let me know if it covers the issue.

pchaigno commented on bpf/lib/nat.h:
```diff
@@ -701,9 +701,15 @@ snat_v4_needs_masquerade(struct __ctx_buff *ctx __maybe_unused,
+			#ifdef ENABLE_REMOTE_NODE_SNAT
+			{
+				target->addr = IPV4_MASQUERADE;
+				return NAT_NEEDED;
```
Did you consider adding a test for this change? I'm a bit worried we regress without one. [A BPF unit test](https://github.com/cilium/cilium/tree/main/bpf/tests) would probably be good enough.

behzad-mir replied:
@pchaigno I understand the concern, but I could not find any test case for snat_v4_needs_masquerade() func (which includes this remote -node-snat mode). I think it's a good suggestion to have a separate PR to cover all test cases for this function including remote node snat.

pchaigno replied:
Fair enough :slightly_smiling_face:

pchaigno replied:
cc @julianwiedmann

tamilmani1989 commented on pkg/datapath/linux/config/config.go:
```diff
@@ -643,6 +643,9 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 		cDefinesMap["SNAT_COLLISION_RETRIES"] = fmt.Sprintf("%d", nat.SnatCollisionRetries)
 
 		if option.Config.EnableBPFMasquerade {
+			if option.Config.EnableRemoteNodeSnat {
```
should we log a warning if someone enables remote node snat without enabling bpf masquerade?

pchaigno replied:
Good point. I'd even just fatal in that case as that's not a config that makes sense. Without a signal, users might be confused as to why the flag is without effect on iptables-based masquerading.

(FYI, such an error should probably go in `daemon_main.go`; we have multiple such error cases.)

behzad-mir replied:
Will add a warning in that case

tamilmani1989 commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -172,6 +172,7 @@ cilium-agent [flags]
       --enable-policy string                                      Enable policy enforcement (default "default")
       --enable-policy-secrets-sync                                Enables Envoy secret sync for Secrets used in CiliumNetworkPolicy and CiliumClusterwideNetworkPolicy
       --enable-recorder                                           Enable BPF datapath pcap recorder
+      --enable-remote-node-snat                                   SNAT IPv4 packets from endpoints leaving the host destined to a remote node
```
can we also mention it requires bpf-masquerade to be enabled.

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -723,8 +729,15 @@ snat_v4_needs_masquerade(struct __ctx_buff *ctx __maybe_unused,
+			}
+			#endif
 			return NAT_PUNT_TO_STACK;
+		}
```
How about adding the `ENABLE_REMOTE_NODE_SNAT` block just once, at the top of the whole section?

Are you planning to add the same handling in the IPv6 path?

behzad-mir replied:
I did not want to make change for the Tunnel mode case that is already in place. For IPv6 We probably need to introduce a new config filed since at least for Azure we are not NAtting traffic to IPv6 in dualstack mode.

julianwiedmann replied:
I see. I think you'll want to describe in a lot more detail what this new feature flag is doing exactly, and in which scenarios it is useful / ineffective. `Documentation/network/concepts/masquerading.rst` is probably a good resting place.

behzad-mir replied:
Will do

pchaigno replied:
> For IPv6 We probably need to introduce a new config filed since at least for Azure we are not NAtting traffic to IPv6 in dualstack mode.

I don't understand how that matters to the new flag. We already have `enableIPv6Masquerade`, so if you don't want to NAT traffic to IPv6, then that can be disabled. It doesn't need to be tied in any way to how remote nodes are treated, does it?

behzad-mir replied:
@julianwiedmann 
Per @pchaigno comment, I actually made use of the same flag to SNAT traffic to IPv6 as well since we already have V6Masqurade in configmap. I also took the condition out of Tunnel mode. Since if this flag is set it wil be an explicit request for masqurading and will trump the tunneling mode condition.
I made a new commeit to address this

tamilmani1989 replied:
> It doesn't need to be tied in any way to how remote nodes are treated, does it?

Agree. There shouldn't be any difference with respect to ipv4/ipv6. the new flag should be common for both.

behzad-mir replied:
The issue has been addressed.

thorn3r reviewed (approved):
agent changes lgtm

julianwiedmann commented on daemon/cmd/daemon_main.go:
```diff
@@ -663,6 +663,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.Bool(option.EnableIPv4Masquerade, true, "Masquerade IPv4 traffic from endpoints leaving the host")
 	option.BindEnv(vp, option.EnableIPv4Masquerade)
 
+	flags.Bool(option.EnableRemoteNodeSnat, false, "SNAT IPv4 packets from endpoints leaving the host destined to a remote node node in BPF routing mode. This option requires to set enable-bpf-masquerade to true.")
```
```suggestion
	flags.Bool(option.EnableRemoteNodeMasquerade, false, "Masquerade packets from endpoints leaving the host destined to a remote node in BPF routing mode. This option requires to set enable-bpf-masquerade to true.")
```

Could you explain what you mean by "BPF routing mode" ? Which config option does that correspond to?

julianwiedmann replied:
Note that instead of hardcoding the `enable-bpf-masquerade` name, you can do a `fmt.Sprintf()` and refer to `option.EnableBPFMasquerade`. See other examples in the file.

julianwiedmann commented on pkg/option/config.go:
```diff
@@ -340,6 +340,9 @@ const (
 	LogOpt = "log-opt"
 
+	// EnableRemoteNodeSnat SNAT IPv4 packets from endpoints leaving the host destined to a remote node node in BPF routing mode. This option requires to set enable-bpf-masquerade to true.
+	EnableRemoteNodeSnat = "enable-remote-node-snat"
```
```suggestion
	// EnableRemoteNodeMasquerade Masquerade packets from endpoints leaving the host destined to a remote node node in BPF routing mode. This option requires to set enable-bpf-masquerade to true.
	EnableRemoteNodeMasquerade = "enable-remote-node-masquerade"
```

Sorry for being pedantic, but I'd really like us to use "masquerading" in this context, so that it matches the other option names.

julianwiedmann commented on Documentation/network/concepts/masquerading.rst:
```diff
@@ -178,3 +178,15 @@ by default, use interfaces listed in the ``devices`` field as the egress masquer
+
+However, this flag does not provide the optimal behavior because packets directed to the nodes' InternalIP addresses will also be masqueraded. 
+To achieve the optimal behavior, we would need to inform the datapath about the type of IP address each node IP address represents. 
+This could potentially be accomplished by extending the node ID map.
```
```suggestion
Masquerade traffic to Remote Nodes
****************************

To masquerade traffic to remote nodes in BPF routing mode, use the option ``enable-remote-node-masquerade: "true"``. 
This option requires ``enable-bpf-masquerade: "true"`` and also either 
``enable-ipv4-masquerade: "true"`` or ``enable-ipv6-masquerade: "true"`` to SNAT traffic for IPv4 and IPv6, respectively.

However, this flag does not provide the optimal behavior because packets directed to the nodes' InternalIP addresses will also be masqueraded. 
To achieve the optimal behavior, we would need to inform the datapath about the type of IP address each node IP address represents. 
This could potentially be accomplished by extending the node ID map.
```

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -707,10 +707,18 @@ snat_v4_needs_masquerade(struct __ctx_buff *ctx __maybe_unused,
 		/* Don't masquerade in native-routing mode: */
 		if (!is_defined(TUNNEL_MODE))
 			return NAT_PUNT_TO_STACK;
-
```
accidental left over?

julianwiedmann reviewed (changes requested):
Thank you! A few clarifications / suggestions inline, I'm mostly unsure what you mean by "BPF routing mode".

Have you thought of how this new feature can be tested, so that folks don't drive by and accidentally break it? Maybe a BPF test in `bpf/tests/` that sets up the needed ipcache entry, and confirms that packets are modified as expected when the feature is enabled?

behzad-mir commented on Documentation/network/concepts/masquerading.rst:
```diff
@@ -178,3 +178,15 @@ by default, use interfaces listed in the ``devices`` field as the egress masquer
 it takes precedence over ``devices`` to choose which network interface should perform masquerading.
 You can set ``egress-masquerade-interfaces`` to match multiple interfaces like this: ``eth+ ens+``.
+
+SNAT traffic to Remote Node
```
Addressed in the new commit.

behzad-mir commented on Documentation/network/concepts/masquerading.rst:
```diff
@@ -181,3 +192,14 @@ by default, use interfaces listed in the ``devices`` field as the egress masquer
 it takes precedence over ``devices`` to choose which network interface should perform masquerading.
 You can set ``egress-masquerade-interfaces`` to match multiple interfaces like this: ``eth+ ens+``.
+
+Masquerade traffic to Remote Nodes
```
resolved in the new commit

julianwiedmann commented on daemon/cmd/daemon_main.go:
```diff
@@ -1254,6 +1257,10 @@ func initEnv(vp *viper.Viper) {
+	if option.Config.EnableRemoteNodeMasquerade && !option.Config.EnableBPFMasquerade {
+		log.Fatal("EnableRemoteNodeMasquerade is enabled but EnableBPFMasquerade is not. Ignoring option.")
+	}
+
```
```suggestion
	if option.Config.EnableRemoteNodeMasquerade && !option.Config.EnableBPFMasquerade {
		log.Warn("EnableRemoteNodeMasquerade is enabled but EnableBPFMasquerade is not. Ignoring option.")
	}

```


Right now we `Fatal()`. If we want to ignore the misconfig & continue, then you'd want a `Warn()` instead.

behzad-mir replied:
We have discussed this in another thread and it was originally warn() @tamilmani1989 asked to make it a fatal()

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -738,10 +738,18 @@ snat_v4_needs_masquerade(struct __ctx_buff *ctx __maybe_unused,
+		/* SNAT the packet to remote node if enable-remote-node-masquerade flag is set to true.
+		 * This is a feature flag that can be enabled in BPF masquerading mode.
+		 */
+		#ifdef ENABLE_REMOTE_NODE_MASQUERADE
```
could you please also wire up some build testing (in `bpf/Makefile`) and complexity testing (in `bpf/complexity-tests/.../bpf_host/...`) for this symbol?

behzad-mir replied:
Added both requested tests in the new commit

julianwiedmann reviewed (approved):
thank you, lgtm! I'm skeptical how many special-cases we can realistically maintain in the BPF masq code, but let's at least give this one a try.

A few non-blocking comments inline, please address them if this PR needs another spin. Alternatively in a follow-up PR.

thorn3r reviewed (approved):
agent changes LGTM :+1:

asauber commented on pkg/option/config.go:
```diff
@@ -331,6 +331,9 @@ const (
 	// LogOpt sets log driver options for cilium
 	LogOpt = "log-opt"
 
+	// EnableRemoteNodeMasquerade Masquerade packets from endpoints leaving the host destined to a remote node node in BPF masquerading mode. This option requires to set enable-bpf-masquerade to true.
```
A few typos here (i.e. "node node"). Suggested description:

"Masquerade packets from endpoints leaving the host destined for a remote node in BPF masquerading mode. This option requires that enable-bpf-masquerade is set to true."

behzad-mir replied:
Addressed in the new commit

joestringer reviewed (changes requested):
Thanks for the PR. Please also address the comments posted by other reviewers in the PR above. (The PR will need a rebase against the tip of `main` while you're at it).

tamilmani1989 commented on bpf/tests/bpf_nat_tests.c:
```diff
@@ -1303,4 +1308,68 @@ int test_nat4_port_allocation_udp_check(struct __ctx_buff *ctx)
+
+	/* Clean up */
+	remote_endpoint_mock = NULL;
+
```
can we add one case for tunnel mode enabled and whats the expected value to be returned by `snat_v4_needs_masquerade`

behzad-mir replied:
Addressed in the new commit.

joestringer commented on Documentation/network/concepts/masquerading.rst:
```diff
@@ -44,6 +44,17 @@ Setting the routable CIDR
+
+However, this flag does not provide the optimal behavior because packets directed to the nodes' InternalIP addresses will also be masqueraded. 
+To achieve the optimal behavior, we would need to inform the datapath about the type of IP address each node IP address represents. 
+This could potentially be accomplished by extending the node ID map.
```
This is user documentation describing the concepts, but the text appears to be proposing future changes that may or may not be put into Cilium. I would suggest to make the text clear and specific about the current behavior and leave forward-looking statements for issue discussions.
```suggestion
This flag currently masquerades traffic to node ``InternalIP`` addresses. This may change in future. See :gh-issue:`####` for further discussion on this topic.
```

Note the suggestion above does not specify a specific github issue; you might consider filing a new one for the proposed solution, or alternatively you could reuse one of the existing ones. Depends whether we intend to close those existing issues assuming this PR is accepted.

behzad-mir replied:
Done in the new commit

joestringer commented on Documentation/network/concepts/masquerading.rst:
```diff
@@ -44,6 +44,17 @@ Setting the routable CIDR
+
+To masquerade traffic to remote nodes in BPF masquerading mode, use the option ``enable-remote-node-masquerade: "true"``. 
+This option requires ``enable-bpf-masquerade: "true"`` and also either 
+``enable-ipv4-masquerade: "true"`` or ``enable-ipv6-masquerade: "true"`` to SNAT traffic for IPv4 and IPv6, respectively.
```
Suggestion to clarify which traffic is affected by the option:

```suggestion
``enable-ipv4-masquerade: "true"`` or ``enable-ipv6-masquerade: "true"`` to SNAT traffic for IPv4 and IPv6, respectively.
This option only affects traffic from an Endpoint directed towards the address of a remote node, and not traffic between Endpoints on different nodes.
```

joestringer commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -168,6 +168,8 @@ cilium-agent [flags]
       --enable-pmtu-discovery                                     Enable path MTU discovery to send ICMP fragmentation-needed replies to the client
       --enable-policy string                                      Enable policy enforcement (default "default")
       --enable-policy-secrets-sync                                Enables Envoy secret sync for Secrets used in CiliumNetworkPolicy and CiliumClusterwideNetworkPolicy
+      --enable-recorder                                           Enable BPF datapath pcap recorder
```
Unrelated?
```suggestion
```

behzad-mir replied:
I think this showed uP after a merge conflict.

joestringer replied:
Your PR should not need to add anything from a merge conflict, could you remove it?

joestringer replied:
This line should be removed.

behzad-mir replied:
done

joestringer commented on bpf/tests/bpf_nat_tests.c:
```diff
@@ -1302,4 +1307,107 @@ int test_nat4_port_allocation_udp_check(struct __ctx_buff *ctx)
 	test_finish();
 }
 
-BPF_LICENSE("Dual BSD/GPL");
```
Why is the license declaration being removed?

behzad-mir replied:
Removed by mistake.

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1261,6 +1264,10 @@ func initEnv(logger *slog.Logger, vp *viper.Viper) {
 	}
 
+	if option.Config.EnableRemoteNodeMasquerade && !option.Config.EnableBPFMasquerade {
+		logging.Fatal("EnableRemoteNodeMasquerade is enabled but EnableBPFMasquerade is not. Ignoring option.")
```
If this is fatal, then the options are not being ignored. This is the common pattern but should be expressed more clearly to the user. Also needs to be rebased for `slog` migration in the upstream tree:
```suggestion
		logging.Fatal(logger, "Option " + option.EnableRemoteNodeMasquerade + " requires BPF masquerade to be enabled (" + option.EnableBPFMasquerade + ")")
```

behzad-mir replied:
I rebased it already and made the recommended changes

joestringer commented on Documentation/network/concepts/masquerading.rst:
```diff
@@ -44,6 +44,17 @@ Setting the routable CIDR
+However, this flag does not provide the optimal behavior because packets directed to the nodes' InternalIP addresses will also be masqueraded. 
+To achieve the optimal behavior, we would need to inform the datapath about the type of IP address each node IP address represents. 
+This could potentially be accomplished by extending the node ID map.
+
```
I could imagine this feature leading to confused deputy attacks in an environment, so it's probably best recommended to be disabled by default:

```suggestion

This option can limit Cilium's ability to enforce ingress host firewall policies for traffic from Pods towards Nodes.
If you use this option, consider establishing default deny policies to prevent Endpoints from communicating with Nodes,
and minimize use of policy statements that allow traffic from remote nodes (such as the ``remote-node`` Entity).
If in doubt, disable this option.

```

joestringer commented on bpf/tests/bpf_nat_tests.c:
```diff
@@ -1302,4 +1307,107 @@ int test_nat4_port_allocation_udp_check(struct __ctx_buff *ctx)
+
+    /* Test 3: With tunnel mode enabled, without ENABLE_REMOTE_NODE_MASQUERADE */
+    #undef ENABLE_REMOTE_NODE_MASQUERADE
+    #define TUNNEL_MODE 1
```
Is there a better way to test this than modifying preprocessor macros mid-test? 

I wonder whether tests specific to this mode should just be declared in a dedicated separate test file in order to avoid tainting the test environment for the remainder of the file.

behzad-mir replied:
Moved the entire test to a new file.

joestringer commented on daemon/cmd/daemon_main.go:
Would you mind adding a bit more detail about the use case into the PR description? Currently it just refers to the behavior being different from upstream ip-masq-agent, but no details about a specific situation where this causes problems that might influence a user to use the new option.

behzad-mir replied:
@joestringer Some more details about the use case has been added to both PR description and release note.
@pchaigno Would appreciate it if you review the PR again.

joestringer commented on Documentation/network/concepts/masquerading.rst:
```diff
@@ -44,6 +44,21 @@ Setting the routable CIDR
+``enable-ipv4-masquerade: "true"`` or ``enable-ipv6-masquerade: "true"`` to SNAT traffic for IPv4 and IPv6, respectively.
+This option only affects traffic from an Endpoint directed towards the address of a remote node, and not traffic between Endpoints on different nodes.
+
+This flag currently masquerades traffic to node ``InternalIP`` addresses. This may change in future. See : (:gh-issue:`35823`) and (:gh-issue:`17177`) for further discussion on this topic.
```
Minor nit for formatting:
```suggestion
This flag currently masquerades traffic to node ``InternalIP`` addresses. This may change in future. See :gh-issue:`35823` and :gh-issue:`17177` for further discussion on this topic.
```

behzad-mir replied:
Been addressed in the new commit

joestringer reviewed (changes requested):
Couple of minor additional nits. I'll defer to @cilium/sig-datapath on the datapath implementation / tests review.

behzad-mir committed:
bpf: nat: enabling masqurade for remote node when BPF routing is enabled.

Currently, Cilium ebpf based masquerading does not perform SNAT
on packets destined for a remote node, even if the node CIDR is not listed in the non-masquerade list.
This behavior differs from the upstream ip-masq-agent.

With this change if the new confimap field "enable-remote-node-snat" is set to "true"
then packets destined to remote nodes will be SNATed in bpf routing mode.

The proposed change has been verified in a cilium bpf routing cluster with enable-remote-node-snat = "true".
ICMP packet from a pod to a remote node initied.
The TCP-dump from remote node indicates that the source IPs for incoming packet SNATed to the ip of source node.

Signed-off-by: Behzad Mirkhanzadeh <behzadm@microsoft.com>

ti-mo commented on pkg/datapath/linux/config/config.go:
```diff
@@ -563,6 +563,9 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 
 		if option.Config.EnableBPFMasquerade {
+			if option.Config.EnableRemoteNodeMasquerade {
+				cDefinesMap["ENABLE_REMOTE_NODE_MASQUERADE"] = "1"
```
Since you're gating a small amount of behaviour, please use the new CONFIG()-style configuration mechanism: https://docs.cilium.io/en/latest/contributing/development/datapath_config/#dpconfig.

behzad-mir replied:
@ti-mo Thank you for your review and comment. Since we have tested this option with current config style and the PR needs to be merged ASAP. I s it possible to do this as a follow up PR?

joestringer replied:
For what it's worth, I understand the desire to move this PR quicker given how long it's been out for and if you already tested the logic, but that's not a motivation to reject the feedback from reviewers. As a general rule in Cilium we prefer not to defer technical debt improvements to follow up PRs, as that can easily fall through the gaps and ultimately burden the maintainers with additional work to maintain the code. For full disclosure as well on the timelines, this PR has slipped the v1.18.x release window so it will now be targeted for v1.19.

behzad-mir replied:
@ti-mo @joestringer I have the changes ready to transition to the new CONFIG() style but noticed if I do the transition then the complexity tests and bpf/Makefile chnages are no longer needed since I am removing the ENBALE_REMOTE_NODE_MASQURADE from the code.
Let me know if there is a way to do the complexity test or we can skip it.

ti-mo replied:
There is since today, see https://github.com/cilium/cilium/pull/40367. :)

behzad-mir replied:
@ti-mo 
I added a new commit. Let me know if it address the issue. Also based on the above PR, do I need to make any specific changes to load my config option for the complexity test or is it going to be picked up by the verifier automatically?

pchaigno commented on Documentation/operations/upgrade.rst:
```diff
@@ -330,6 +330,10 @@ communicating via the proxy must reconnect to re-establish connections.
+* ``enable-remote-node-masquerade`` config option is introduced. To masquerade traffic to remote nodes in BPF masquerading mode, use the option ``enable-remote-node-masquerade: "true"``. This option requires ``enable-bpf-masquerade: "true"`` and also either
+  ``enable-ipv4-masquerade: "true"`` or ``enable-ipv6-masquerade: "true"`` to SNAT traffic for IPv4 and IPv6, respectively. This flag currently masquerades traffic to node ``InternalIP`` addresses. This may change in future. See :gh-issue:`35823` and :gh-issue:`17177` for further discussion on this topic.
+  This option can limit Cilium's ability to enforce ingress host firewall policies for traffic from Pods towards Nodes.
+  If you use this option, consider establishing default deny policies to prevent Endpoints from communicating with Nodes, and minimize use of policy statements that allow traffic from remote nodes (such as the ``remote-node`` Entity). If in doubt, disable this option.
```
Let's keep this short, like for surrounding entries. The goal here is to let users know about new flags, not describe them in details.

behzad-mir replied:
@pchaigno I removed the second part of it. Let me know if you prefer to short it even more.

pchaigno replied:
I think it's okay as is.

pchaigno commented on Documentation/operations/upgrade.rst:
```diff
@@ -330,6 +330,8 @@ communicating via the proxy must reconnect to re-establish connections.
   your network policies in advance for the Cilium 1.19 upgrade.
 * Creating or deleting policies via the local REST api is deprecated. This will be removed entirely in v1.19.
+* ``enable-remote-node-masquerade`` config option is introduced. To masquerade traffic to remote nodes in BPF masquerading mode, use the option ``enable-remote-node-masquerade: "true"``. This option requires ``enable-bpf-masquerade: "true"`` and also either
+  ``enable-ipv4-masquerade: "true"`` or ``enable-ipv6-masquerade: "true"`` to SNAT traffic for IPv4 and IPv6, respectively. This flag currently masquerades traffic to node ``InternalIP`` addresses. This may change in future. See :gh-issue:`35823` and :gh-issue:`17177` for further discussion on this topic.
```
Nit: Please follow the line width of other entries.

behzad-mir replied:
Resolved in the new commit. Added this section to 1.19 updates and made the lines shorter.

pchaigno replied:
There must have been a mistake because it doesn't look resolve AFAICT.

behzad-mir replied:
tagged you below in the updated version.

behzad-mir commented on Documentation/operations/upgrade.rst:
```diff
@@ -293,6 +293,12 @@ communicating via the proxy must reconnect to re-establish connections.
+  use the option ``enable-remote-node-masquerade: "true"``. This option requires ``enable-bpf-masquerade: "true"`` and also either
+  ``enable-ipv4-masquerade: "true"`` or ``enable-ipv6-masquerade: "true"`` to SNAT traffic for IPv4 and IPv6, respectively. 
+  This flag currently masquerades traffic to node ``InternalIP`` addresses.
+  This may change in future. See :gh-issue:`35823` and :gh-issue:`17177` for further discussion on this topic.
```
@pchaigno This is the updated version in new commit. Let me know if the line width are correct.

joestringer replied:
FWIW it looks like we haven't been super consistent in enforcing line lengths in this file, but for most Cilium contributors we have our editors set up for 80 character lines. Unfortunately we can't really set up a simple linter to ensure this because sometimes links cause lines to be longer, so this tends to be more of a soft limit in the docs.

We often also have linters that highlight unnecessary whitespace at EOL, and it looks like these lines also have some there too.

This incremental patch would fix up both aspects:

```
diff --git a/Documentation/operations/upgrade.rst b/Documentation/operations/upgrade.rst
index 62915fc8659c..372add2b6f8c 100644
--- a/Documentation/operations/upgrade.rst
+++ b/Documentation/operations/upgrade.rst
@@ -293,11 +293,14 @@ communicating via the proxy must reconnect to re-establish connections.
 1.19 Upgrade Notes
 ------------------
 
-* ``enable-remote-node-masquerade`` config option is introduced. To masquerade traffic to remote nodes in BPF masquerading mode, 
-  use the option ``enable-remote-node-masquerade: "true"``. This option requires ``enable-bpf-masquerade: "true"`` and also either
-  ``enable-ipv4-masquerade: "true"`` or ``enable-ipv6-masquerade: "true"`` to SNAT traffic for IPv4 and IPv6, respectively. 
-  This flag currently masquerades traffic to node ``InternalIP`` addresses.
-  This may change in future. See :gh-issue:`35823` and :gh-issue:`17177` for further discussion on this topic.
+* ``enable-remote-node-masquerade`` config option is introduced. To masquerade
+  traffic to remote nodes in BPF masquerading mode, use the option
+  ``enable-remote-node-masquerade: "true"``. This option requires
+  ``enable-bpf-masquerade: "true"`` and also either ``enable-ipv4-masquerade:
+  "true"`` or ``enable-ipv6-masquerade: "true"`` to SNAT traffic for IPv4 and
+  IPv6, respectively. This flag currently masquerades traffic to node
+  ``InternalIP`` addresses. This may change in future. See :gh-issue:`35823`
+  and :gh-issue:`17177` for further discussion on this topic.
 
 Removed Options
 ~~~~~~~~~~~~~~~
```

pchaigno replied:
You'll also need to squash the two commits at the same time. (Or split changes in multiple commits, with ex. functional changes, docs, tests.)

behzad-mir replied:
Yes I just did that @pchaigno .
I left them like that intentionally for review purpose since the second commit was changing the whole config process. I squashed and rebased to the master.

rgo3 commented on bpf/Makefile:
```diff
@@ -93,6 +93,12 @@ HOST_OPTIONS = $(LXC_OPTIONS) \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DPOLICY_VERDICT_NOTIFY:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_PREFILTER:-DENABLE_SESSION_AFFINITY:-DENABLE_HOST_FIREWALL:-DENABLE_ICMP_RULE:-DENABLE_SRV6:-DENABLE_SRV6_SRH_ENCAP:-DENABLE_SCTP:-DENABLE_VTEP:-DENABLE_MULTICAST:-DENCRYPTED_OVERLAY: \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DPOLICY_VERDICT_NOTIFY:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_DSR:-DDSR_ENCAP_MODE:-DDSR_ENCAP_GENEVE:-DENABLE_PREFILTER:-DENABLE_SESSION_AFFINITY:-DENABLE_HOST_FIREWALL:-DENABLE_ICMP_RULE:-DENABLE_SRV6:-DENABLE_SRV6_SRH_ENCAP:-DENCRYPTED_OVERLAY:-DENABLE_SCTP:-DENABLE_VTEP:-DDSR_ENCAP_IPIP=2 \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DPOLICY_VERDICT_NOTIFY:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_PREFILTER:-DENABLE_SESSION_AFFINITY:-DENABLE_HOST_FIREWALL:-DENABLE_ICMP_RULE:-DENABLE_SRV6:-DENABLE_SRV6_SRH_ENCAP:-DENCRYPTED_OVERLAY:-DENABLE_SCTP:-DENABLE_VTEP:-DENABLE_IPSEC: \
+    -DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_BPF_MASQUERADE:-DENABLE_MASQUERADE_IPV4:-DENABLE_MASQUERADE_IPV6: \
```
Now that you've made `enable_remote_node_masquerade` and load time config, are these makefile changes still necessary?

behzad-mir replied:
@rgo3  I removed the changes from the Makefile.

rgo3 commented on pkg/datapath/config/host_config.go:
```diff
@@ -14,6 +14,8 @@ type BPFHost struct {
 	// Pass traffic with extended IP protocols.
 	EnableExtendedIPProtocols bool `config:"enable_extended_ip_protocols"`
+	// Masquerade packets from endpoints destined to a remote node.
+	EnableRemoteNodeMasquerade bool `config:"enable_remote_node_masquerade"`
```
As Timo suggested in his earlier comments, you can add a permutation of this config to the complexity tests of the bpf_host progs, see https://github.com/cilium/cilium/blob/5a60ca4e9c4152cb3ce85276cb6084830200d886/pkg/datapath/loader/verifier_load_test.go#L24.

The following should achieve what you want:
```
func hostLoadPermutations() iter.Seq[*config.BPFHost] {
	return func(yield func(*config.BPFHost) bool) {
		for permutation := range permute(2) {
			cfg := config.NewBPFHost(*config.NewNode())
			cfg.SecctxFromIPCache = permutation[0]
            cfg.EnableRemoteNodeMasquerade = permutation[1]
			if !yield(cfg) {
				return
			}
		}
	}
}
```

behzad-mir replied:
Thanks for pointing this out I added cfg.EnableRemoteNodeMasquerade = permutation[1] to the exiting hostLoadPermutations() function in the file.

behzad-mir replied:
@rgo3 let me know if the changes address the issue.

rgo3 replied:
Looks good!

rgo3 reviewed (changes requested):
Picking up the review from Timo here, as I was assigned initially for loader anyway. Overall looks good, we should just make sure the config variable is covered in verifier tests.

rgo3 commented on pkg/datapath/loader/verifier_load_test.go:
```diff
@@ -26,6 +26,7 @@ func hostLoadPermutations() iter.Seq[*config.BPFHost] {
 		for permutation := range permute(1) {
```
```suggestion
		for permutation := range permute(2) {
```

I think you'll need to do this since your are now permuting two boolean values. Accessing permutation[1] without this change should also be index out of bounds if you'd run this test locally.

behzad-mir replied:
corrected.

rgo3 commented on pkg/datapath/loader/loader.go:
```diff
@@ -203,6 +203,10 @@ func netdevRewrites(ep datapath.EndpointConfiguration, lnc *datapath.LocalNodeCo
 			cfg.NATIPv6Masquerade = ipv6.As16()
 		}
+		// Masquerading IPv4 traffic from endpoints leaving the host.
+		if option.Config.EnableRemoteNodeMasquerade {
```
Nit
```suggestion
		cfg.EnableRemoteNodeMasquerade = option.Config.EnableRemoteNodeMasquerade
```

rgo3 reviewed (approved):
Loader changes LGTM!
